Expose syncError message from ArgoCD Application to GitopsDeploymentS…

Rizwana777 opened the pull request:
Description:

 - Expose SyncError field of ArgoCD Application GitopsDeploymentStatus field
 - Added a new field called 'sync_error' in database
 - Added new field called SyncError in GitopsDeploymentStatus field
 - Retrieve the sync_error field from database and expose it to SyncError field of GitopsDeploymentStatus
 - Added unit test cases for SyncError field

Link to JIRA Story (if applicable):
https://issues.redhat.com/browse/GITOPSRVCE-203

chetan-rns commented on cluster-agent/controllers/argoproj.io/application_controller_test.go:
```diff
@@ -566,6 +582,229 @@ var _ = Describe("Application Controller", func() {
+				Health:                          "Healthy",
+				Sync_Status:                     "Synced",
+				ReconciledState:                 reconciledStateString,
+				SyncError:                       "test-sync-error",
```
We are directly updating the sync error message in the DB object. Shouldn't we update it in the Argo CD Application?

Rizwana777 replied:
Here the test is as follows, where  SyncError is already present in db  which is "test-sync-error" and once reconcile function is called we need to check whether db is updated with the syncError similar to the ArgoCD Application syncError

chetan-rns commented on tests-e2e/fixture/application/fixture.go:
```diff
@@ -124,3 +125,36 @@ func HaveSyncStatusCode(status appv1alpha1.ApplicationStatus) matcher.GomegaMatc
+		var res bool
+
+		for _, msg := range app.Status.Conditions {
+			existingConditionMessage = msg.Message
```
Should we check if the condition is of type "SyncError" before copying the message? There could be other conditions as well.

Rizwana777 commented on cluster-agent/controllers/argoproj.io/application_controller.go:
```diff
@@ -149,6 +149,14 @@ func (r *ApplicationReconciler) Reconcile(ctx context.Context, req ctrl.Request)
+			// Update applicationState.SyncError with the ArgoCD application syncError message
+			for _, syncError := range app.Status.Conditions {
+				// Update syncError field of appliactionState only if type is SyncError
+				if syncError.Type == "SyncError" {
```
"SyncError" is used only once here so I have not created constant, I will create one if its need in future

Rizwana777 commented on utilities/db-migration/migrations/000007_v7.up.sql:
```diff
@@ -1 +1,2 @@
 ALTER TABLE ApplicationState ADD COLUMN reconciled_state VARCHAR ( 4096 );
+ALTER TABLE ApplicationState ADD COLUMN sync_error VARCHAR ( 4096 );
```
I was not sure about this, as migration for ApplicationState already present so added it in the same file, please let me know if I am wrong

Rizwana777 commented on cluster-agent/controllers/argoproj.io/application_controller_test.go:
```diff
@@ -566,6 +582,229 @@ var _ = Describe("Application Controller", func() {
+
+			ctx = context.Background()
+
+			guestbookApp := &appv1.Application{
```
no @chetan-rns here condition is different, test is for checking if type is not equal to syncError, then we need to check  whether syncError field is empty

Rizwana777 commented on cluster-agent/controllers/argoproj.io/application_controller_test.go:
```diff
@@ -417,6 +426,13 @@ var _ = Describe("Application Controller", func() {
 				},
 			}
 
+			app := appv1.ApplicationCondition{
```
yes removing it

Rizwana777 commented:
@chetan-rns Addressed all the comments , PTAL

jgwest commented on cluster-agent/controllers/argoproj.io/application_controller.go:
```diff
@@ -149,6 +149,14 @@ func (r *ApplicationReconciler) Reconcile(ctx context.Context, req ctrl.Request)
+			for _, syncError := range app.Status.Conditions {
+				// Update syncError field of appliactionState only if type is SyncError
+				if syncError.Type == appv1.ApplicationConditionSyncError {
+					applicationState.SyncError = db.TruncateVarchar(syncError.Message, db.ApplicationStateSyncErrorLength)
```
This logic should also handle the case where the ApplicationConditionSyncError is not found, in which case SyncError should be set to `""`

Rizwana777 replied:
yes right @jgwest  I didn't realised this 😅, thank you , I will update this

jgwest commented on cluster-agent/controllers/argoproj.io/application_controller_test.go:
```diff
@@ -566,6 +575,229 @@ var _ = Describe("Application Controller", func() {
+			err = reconciler.DB.CreateApplication(ctx, applicationDB)
+			Expect(err).To(BeNil())
+
+			By("Verify SyncError is not present in database")
```
This `By` message looks like it's not describing what the step is doing.

jgwest commented on cluster-agent/controllers/argoproj.io/application_controller_test.go:
```diff
@@ -566,6 +575,229 @@ var _ = Describe("Application Controller", func() {
+
+			// Checking for new values in database object
+			for _, syncError := range guestbookApp.Status.Conditions {
+				Expect(applicationStateDB.SyncError).To(Equal(syncError.Message))
```
Can change this comment to `By("verifying that the ApplicationState DB row has been updated to match the Application SyncError")`

jgwest commented on cluster-agent/controllers/argoproj.io/application_controller_test.go:
```diff
@@ -566,6 +575,229 @@ var _ = Describe("Application Controller", func() {
+					},
+					Conditions: []appv1.ApplicationCondition{
+						{
+							Type:               "test-error",
```
Rather than using `test-error`, you can use one of the other error types, like `ApplicationConditionUnknownError`

jgwest commented on cluster-agent/controllers/argoproj.io/application_controller_test.go:
```diff
@@ -566,6 +575,229 @@ var _ = Describe("Application Controller", func() {
+			err = reconciler.DB.CreateApplication(ctx, applicationDB)
+			Expect(err).To(BeNil())
+
+			By("Verify SyncError is not present in database")
```
As above, by message seems off here.

jgwest commented on cluster-agent/controllers/argoproj.io/application_controller_test.go:
```diff
@@ -566,6 +575,229 @@ var _ = Describe("Application Controller", func() {
+
+			// Checking for syncError value in database object
+			for _, syncError := range guestbookApp.Status.Conditions {
+				Expect(applicationStateDB.SyncError).To(BeEmpty())
```
I think this for loop can just be replaced with `Expect(applicationStateDB.SyncError).To(BeEmpty())`?

jgwest commented on cluster-agent/controllers/argoproj.io/application_controller_test.go:
```diff
@@ -566,6 +575,229 @@ var _ = Describe("Application Controller", func() {
+			// Checking for new values in database object
+			for _, syncError := range guestbookApp.Status.Conditions {
+				Expect(applicationStateDB.SyncError).To(Equal(syncError.Message))
+				Expect(syncError.Type).To(Equal("SyncError"))
```
I think this for loop can just be replaced with `Expect(applicationStateDB.SyncError).To(Equal("Failed to sync"))`

jgwest commented on cluster-agent/controllers/argoproj.io/application_controller_test.go:
```diff
@@ -566,6 +575,229 @@ var _ = Describe("Application Controller", func() {
+			Expect(err).To(BeNil())
+
+			// Checking for new values in database object
+			for _, syncError := range guestbookApp.Status.Conditions {
```
Likewise, this for loop can be replaced, as above.

jgwest reviewed (changes requested):
Looks good @Rizwana777, I've made some requested changes to check out: the main item is that we should use the `.status.conditions` of `GitOpsDeployment` to store the `SyncError`, rather than creating a new field (more details in the comment).

This should be a quick fix (along with updating the E2E test and E2E test fixture), then it should be good to go.

Rizwana777 commented:
@jgwest addresses all the comments, PTAL, let me know if I missed anything, I am inserting syncError in status.Conditions now, I am not sure whether I have followed the right way to do this, please let me know the corrections if I have followed wrong way, thank you

jgwest reviewed (changes requested):
Looks great @Rizwana777! A couple more tiny tweaks than good to go.

Rizwana777 commented on backend/eventloop/application_event_loop/application_event_runner_test.go:
```diff
@@ -837,6 +839,36 @@ var _ = Describe("ApplicationEventLoop Test", func() {
 			Expect(gitopsDeployment.Status.ReconciledState.Source.RepoURL).To(Equal(reconciledobj.Source.RepoURL))
 			Expect(gitopsDeployment.Status.ReconciledState.Source.Branch).To(Equal(reconciledobj.Source.TargetRevision))
 			Expect(gitopsDeployment.Status.ReconciledState.Destination.Namespace).To(Equal(reconciledobj.Destination.Namespace))
+			Expect(gitopsDeployment.Status.Conditions[0].Message).To(Equal(applicationState.SyncError))
```
yes, I will update this

Rizwana777 commented:
/retest

Rizwana777 commented:
changes updated @jgwest PTAL

jgwest reviewed (approved):
LGTM, thanks @Rizwana777, great that we now get this information from Argo CD
